File: substrate/antiek_bench_weekly_usage_learn_compose.py
"""Antiek-bench weekly usage-learn compose (pure).

backlog_mutated and store_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

UsageOutcome = Literal["worked", "failed", "mixed", "unknown"]
VALID_OUTCOME = frozenset(("worked", "failed", "mixed", "unknown"))
Emphasis = Literal[
    "expand_failure_cases", "expand_success_cases", "hold_stable"
]


class AntiekBenchWeeklyUsageLearnComposeError(ValueError):
    """Fail-closed validation for weekly usage learn."""


@dataclass(frozen=True)
class SubBenchmarkRewriteProposal:
    task: str
    reason: str
    emphasis: Emphasis
    event_count: int
    failed_count: int
    worked_count: int

    def to_dict(self) -> dict[str, Any]:
        return {
            "task": self.task,
            "reason": self.reason,
            "emphasis": self.emphasis,
            "event_count": self.event_count,
            "failed_count": self.failed_count,
            "worked_count": self.worked_count,
        }


@dataclass(frozen=True)
class AntiekBenchWeeklyUsageLearnCompose:
    week_id: str
    event_count: int
    task_count: int
    proposals: tuple[SubBenchmarkRewriteProposal, ...]
    proposal_count: int
    learn_ready: bool
    backlog_mutated: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "week_id": self.week_id,
            "event_count": self.event_count,
            "task_count": self.task_count,
            "proposals": [p.to_dict() for p in self.proposals],
            "proposal_count": self.proposal_count,
            "learn_ready": self.learn_ready,
            "backlog_mutated": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": "antiek_bench_weekly_usage_learn_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise AntiekBenchWeeklyUsageLearnComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_antiek_bench_weekly_usage_learn(
    *,
    week_id: object,
    events: object,
    operator_ack: object,
    min_events_per_task: object | None = None,
) -> AntiekBenchWeeklyUsageLearnCompose:
    """Propose sub-benchmark rewrites from weekly usage. Never mutates store."""
    if not isinstance(operator_ack, bool):
        raise AntiekBenchWeeklyUsageLearnComposeError(
            "operator_ack must be an explicit boolean"
        )
    wid = _require_nonempty(week_id, field="week_id")
    if not isinstance(events, list):
        raise AntiekBenchWeeklyUsageLearnComposeError("events must be an array")

    min_events = 3 if min_events_per_task is None else min_events_per_task
    if not isinstance(min_events, int) or isinstance(min_events, bool) or min_events < 1:
        raise AntiekBenchWeeklyUsageLearnComposeError(
            "min_events_per_task must be a positive integer"
        )

    notes: list[str] = [
        "backlog_mutated=false — rewrite proposals are advisory only",
        "store_mutated=false — Antiek-bench store not written",
        "usage events are caller-supplied only (no invent)",
    ]

    by_task: dict[str, dict[str, int]] = {}
    seen: set[str] = set()
    for i, e in enumerate(events):
        if not isinstance(e, dict):
            raise AntiekBenchWeeklyUsageLearnComposeError(
                f"events[{i}] must be an object"
            )
        eid = _require_nonempty(e.get("event_id"), field=f"events[{i}].event_id")
        if eid in seen:
            raise AntiekBenchWeeklyUsageLearnComposeError(
                f"duplicate event_id: {eid}"
            )
        seen.add(eid)
        task = _require_nonempty(e.get("task"), field=f"events[{i}].task")
        _require_nonempty(e.get("model_id"), field=f"events[{i}].model_id")
        outcome = e.get("outcome")
        if outcome not in VALID_OUTCOME:
            raise AntiekBenchWeeklyUsageLearnComposeError(
                f"events[{i}].outcome must be worked|failed|mixed|unknown"
            )
        score = e.get("score")
        if score is not None:
            if not isinstance(score, (int, float)) or isinstance(score, bool):
                raise AntiekBenchWeeklyUsageLearnComposeError(
                    f"events[{i}].score must be finite in [0,1] when set"
                )
            sf = float(score)
            if sf != sf or sf < 0 or sf > 1:
                raise AntiekBenchWeeklyUsageLearnComposeError(
                    f"events[{i}].score must be finite in [0,1] when set"
                )
        agg = by_task.setdefault(
            task, {"worked": 0, "failed": 0, "mixed": 0, "unknown": 0, "count": 0}
        )
        agg["count"] += 1
        agg[str(outcome)] += 1

    event_count = len(events)
    task_count = len(by_task)
    notes.append(f"event_count={event_count} · task_count={task_count}")

    proposals: list[SubBenchmarkRewriteProposal] = []
    for task, agg in by_task.items():
        if agg["count"] < min_events:
            notes.append(
                f"task={task} skipped (events={agg['count']} < min={min_events})"
            )
            continue
        worked = agg["worked"]
        failed = agg["failed"]
        count = agg["count"]
        if failed > worked:
            emphasis: Emphasis = "expand_failure_cases"
            reason = f"failed={failed} > worked={worked} over {count} events"
        elif worked > failed and worked >= (count + 1) // 2:
            emphasis = "hold_stable"
            reason = f"worked={worked} dominates over {count} events"
        elif worked > 0 and failed == 0:
            emphasis = "expand_success_cases"
            reason = f"all non-fail outcomes worked={worked} over {count} events"
        else:
            emphasis = "expand_failure_cases"
            reason = (
                f"mixed/unknown heavy over {count} events "
                f"(failed={failed} worked={worked})"
            )
        proposals.append(
            SubBenchmarkRewriteProposal(
                task=task,
                reason=reason,
                emphasis=emphasis,
                event_count=count,
                failed_count=failed,
                worked_count=worked,
            )
        )

    proposals.sort(key=lambda p: p.task)
    proposal_count = len(proposals)
    notes.append(f"proposal_count={proposal_count}")

    learn_ready = operator_ack and proposal_count >= 1
    if not operator_ack:
        notes.append("learn_ready=false — operator_ack required")
    elif proposal_count == 0:
        notes.append(
            "learn_ready=false — no tasks met min_events threshold (no invent proposals)"
        )
    else:
        notes.append(
            "learn_ready=true — advisory rewrite proposals ready for operator review"
        )
    notes.extend(("backlog_mutated=false", "store_mutated=false"))

    return AntiekBenchWeeklyUsageLearnCompose(
        week_id=wid,
        event_count=event_count,
        task_count=task_count,
        proposals=tuple(proposals),
        proposal_count=proposal_count,
        learn_ready=learn_ready,
        backlog_mutated=False,
        store_mutated=False,
        notes=tuple(notes),
        authority="antiek_bench_weekly_usage_learn_compose_advisory",
    )


__all__ = [
    "AntiekBenchWeeklyUsageLearnCompose",
    "AntiekBenchWeeklyUsageLearnComposeError",
    "SubBenchmarkRewriteProposal",
    "compose_antiek_bench_weekly_usage_learn",
]

File: apps/reading/src/api/antiekBenchWeeklyUsageLearnCompose.ts
/**
 * Antiek-bench weekly usage-learn compose (pure).
 *
 * Operator vision: recursive benchmark that learns from usage patterns —
 * what worked / didn't in a week — to re-write the benchmark and
 * differentiating sub-benchmarks as the platform expands.
 *
 * This pure layer proposes rewrite recommendations only.
 * backlog_mutated always false.
 * store_mutated always false.
 * Never invents usage events.
 */

export type UsageOutcome = "worked" | "failed" | "mixed" | "unknown";

export interface WeeklyUsageEvent {
  event_id: string;
  /** Task family (deep_research, twin_notes, etc.). */
  task: string;
  model_id: string;
  outcome: UsageOutcome;
  /** Optional quality score 0..1. */
  score?: number | null;
}

export interface SubBenchmarkRewriteProposal {
  task: string;
  /** Why rewrite is proposed — derived from counts, not invented prose. */
  reason: string;
  /** Suggested emphasis: more cases for failures, keep stable for wins. */
  emphasis: "expand_failure_cases" | "expand_success_cases" | "hold_stable";
  event_count: number;
  failed_count: number;
  worked_count: number;
}

export interface AntiekBenchWeeklyUsageLearnInput {
  week_id: string;
  events: WeeklyUsageEvent[];
  operator_ack: boolean;
  /**
   * Minimum events per task before proposing a rewrite.
   * Default 3 when omitted.
   */
  min_events_per_task?: number;
}

export interface AntiekBenchWeeklyUsageLearnCompose {
  week_id: string;
  event_count: number;
  task_count: number;
  proposals: SubBenchmarkRewriteProposal[];
  proposal_count: number;
  /** True when ≥1 proposal and operator_ack. */
  learn_ready: boolean;
  /** Always false — pure layer never mutates bench backlog. */
  backlog_mutated: false;
  /** Always false — pure layer never writes bench store. */
  store_mutated: false;
  notes: string[];
  authority: "antiek_bench_weekly_usage_learn_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

const VALID_OUTCOME = new Set<UsageOutcome>([
  "worked",
  "failed",
  "mixed",
  "unknown",
]);

/**
 * Propose Antiek-bench sub-benchmark rewrites from weekly usage events.
 * Never invents events; never mutates bench store/backlog.
 */
export function composeAntiekBenchWeeklyUsageLearn(
  input: AntiekBenchWeeklyUsageLearnInput,
): AntiekBenchWeeklyUsageLearnCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const week_id = requireNonEmpty(input.week_id, "week_id");
  if (!Array.isArray(input.events)) {
    throw new Error("events must be an array");
  }

  const minEvents =
    input.min_events_per_task === undefined ||
    input.min_events_per_task === null
      ? 3
      : input.min_events_per_task;
  if (
    typeof minEvents !== "number" ||
    !Number.isInteger(minEvents) ||
    minEvents < 1
  ) {
    throw new Error("min_events_per_task must be a positive integer");
  }

  const notes: string[] = [
    "backlog_mutated=false — rewrite proposals are advisory only",
    "store_mutated=false — Antiek-bench store not written",
    "usage events are caller-supplied only (no invent)",
  ];

  type Agg = {
    worked: number;
    failed: number;
    mixed: number;
    unknown: number;
    count: number;
  };
  const byTask = new Map<string, Agg>();
  const seen = new Set<string>();

  for (let i = 0; i < input.events.length; i++) {
    const e = input.events[i];
    if (!e || typeof e !== "object") {
      throw new Error(`events[${i}] must be an object`);
    }
    const eid = requireNonEmpty(e.event_id, `events[${i}].event_id`);
    if (seen.has(eid)) {
      throw new Error(`duplicate event_id: ${eid}`);
    }
    seen.add(eid);
    const task = requireNonEmpty(e.task, `events[${i}].task`);
    requireNonEmpty(e.model_id, `events[${i}].model_id`);
    if (!VALID_OUTCOME.has(e.outcome)) {
      throw new Error(
        `events[${i}].outcome must be worked|failed|mixed|unknown`,
      );
    }
    if (e.score !== undefined && e.score !== null) {
      if (
        typeof e.score !== "number" ||
        !Number.isFinite(e.score) ||
        e.score < 0 ||
        e.score > 1
      ) {
        throw new Error(
          `events[${i}].score must be finite in [0,1] when set`,
        );
      }
    }
    let agg = byTask.get(task);
    if (!agg) {
      agg = { worked: 0, failed: 0, mixed: 0, unknown: 0, count: 0 };
      byTask.set(task, agg);
    }
    agg.count += 1;
    if (e.outcome === "worked") agg.worked += 1;
    else if (e.outcome === "failed") agg.failed += 1;
    else if (e.outcome === "mixed") agg.mixed += 1;
    else agg.unknown += 1;
  }

  const event_count = input.events.length;
  const task_count = byTask.size;
  notes.push(`event_count=${event_count} · task_count=${task_count}`);

  const proposals: SubBenchmarkRewriteProposal[] = [];
  for (const [task, agg] of byTask) {
    if (agg.count < minEvents) {
      notes.push(
        `task=${task} skipped (events=${agg.count} < min=${minEvents})`,
      );
      continue;
    }
    let emphasis: SubBenchmarkRewriteProposal["emphasis"];
    let reason: string;
    if (agg.failed > agg.worked) {
      emphasis = "expand_failure_cases";
      reason = `failed=${agg.failed} > worked=${agg.worked} over ${agg.count} events`;
    } else if (agg.worked > agg.failed && agg.worked >= Math.ceil(agg.count / 2)) {
      emphasis = "hold_stable";
      reason = `worked=${agg.worked} dominates over ${agg.count} events`;
    } else if (agg.worked > 0 && agg.failed === 0) {
      emphasis = "expand_success_cases";
      reason = `all non-fail outcomes worked=${agg.worked} over ${agg.count} events`;
    } else {
      emphasis = "expand_failure_cases";
      reason = `mixed/unknown heavy over ${agg.count} events (failed=${agg.failed} worked=${agg.worked})`;
    }
    proposals.push({
      task,
      reason,
      emphasis,
      event_count: agg.count,
      failed_count: agg.failed,
      worked_count: agg.worked,
    });
  }

  // Stable order by task id
  proposals.sort((a, b) => a.task.localeCompare(b.task));
  const proposal_count = proposals.length;
  notes.push(`proposal_count=${proposal_count}`);

  const learn_ready = input.operator_ack && proposal_count >= 1;
  if (!input.operator_ack) {
    notes.push("learn_ready=false — operator_ack required");
  } else if (proposal_count === 0) {
    notes.push(
      "learn_ready=false — no tasks met min_events threshold (no invent proposals)",
    );
  } else {
    notes.push(
      "learn_ready=true — advisory rewrite proposals ready for operator review",
    );
  }

  notes.push("backlog_mutated=false");
  notes.push("store_mutated=false");

  return {
    week_id,
    event_count,
    task_count,
    proposals,
    proposal_count,
    learn_ready,
    backlog_mutated: false,
    store_mutated: false,
    notes,
    authority: "antiek_bench_weekly_usage_learn_compose_advisory",
  };
}

export function formatAntiekBenchWeeklyUsageLearnSummary(
  c: AntiekBenchWeeklyUsageLearnCompose,
): string {
  return (
    `learn_ready=${c.learn_ready} · proposals=${c.proposal_count} · ` +
    `events=${c.event_count} · backlog_mutated=false · store_mutated=false`
  );
}
